exla/mix.exs

```diff
@@ -114,17 +115,53 @@ defmodule EXLA.MixProject do
 
       _ ->
         File.rm_rf!(xla_extension_path)
-
         Mix.shell().info("Unpacking #{xla_archive_path} into #{cache_dir}")
 
         case :erl_tar.extract(xla_archive_path, [:compressed, cwd: cache_dir]) do
-          :ok -> :ok
+          :ok -> File.write!(xla_snapshot_path, xla_archive_path)
           {:error, term} -> Mix.raise("failed to extract xla archive, reason: #{inspect(term)}")
         end
-
-        File.write!(xla_snapshot_path, xla_archive_path)
     end
 
     {:ok, []}
   end
+
+  defp cached_make(_) do
+    contents =
+      for path <- Path.wildcard("c_src/**/*"),
+          {:ok, contents} <- [File.read(path)],
+          do: contents
+
+    md5 = [XLA.archive_path!() | contents] |> :erlang.md5() |> Base.encode16()
+
+    cache_key =
+      "elixir-#{System.version()}-erts-#{:erlang.system_info(:version)}-xla-#{Application.spec(:xla, :vsn)}-exla-#{@version}-#{md5}"
+
+    cached_so = Path.join([xla_cache_dir(), "exla", cache_key, "libexla.so"])
+    cached? = File.exists?(cached_so)
+
+    if cached? do
+      Mix.shell().info("Using libexla.so from #{cached_so}")
+      File.cp!(cached_so, "cache/libexla.so")
+    end
+
+    result = Mix.Tasks.Compile.ElixirMake.run([])
+
+    if not cached? and match?({:ok, _}, result) do
+      Mix.shell().info("Caching libexla.so at #{cached_so}")
+      File.mkdir_p!(Path.dirname(cached_so))
+      File.cp!("cache/libexla.so", cached_so)
+    end
+
+    result
+  end
+
+  defp xla_cache_dir() do
+    # The directory where we store all the archives
+    if dir = System.get_env("XLA_CACHE_DIR") do
+      Path.expand(dir)
+    else
+      :filename.basedir(:user_cache, "xla")
+    end
+  end
 end
```

Comment:
Should we make it a public function in XLA?

Reply:
Nah, because it doesn't even have to be same dir. I am storing together just to ease deletion.